### ISSUE TITLE
Grunt fails in production mode

### ISSUE THREAD

**Issue by wyqydsyq** [balderdashy/sails#2441]
When I set my environment to 'production', I get the following error when attempting `sails lift`:

````bash
data:    app.js:8674 - ** Grunt :: An error occurred. **
data:    app.js:8674 - error: 
data:    app.js:8674 - ------------------------------------------------------------------------
data:    app.js:8674 - Aborted due to warnings.
data:    app.js:8674 - Running "uglify:dist" (uglify) task
data:    app.js:8674 - { message: 'Name expected',
data:    app.js:8674 -   line: 5976,
data:    app.js:8674 -   col: 3,
data:    app.js:8674 -   pos: 338160,
data:    app.js:8674 -   stack: 'Error\n    at new JS_Parse_Error (/home/damon/bin/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:189:18)\n    at js_error (/home/damon/bin/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:197:11)\n    at croak (/home/damon/bin/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:660:41)\n    at as_symbol (/home/damon/bin/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:1275:27)\n    at vardefs (/home/damon/bin/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:1056:25)\n    at var_ (/home/damon/bin/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:1070:27)\n    at /home/damon/bin/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:818:30\n    at /home/damon/bin/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:703:24\n    at block_ (/home/damon/bin/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:977:20)\n    at /home/damon/bin/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:950:25\n    at function_ (/home/damon/bin/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:955:15)\n    at expr_atom (/home/damon/bin/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:1158:24)\n    at maybe_unary (/home/damon/bin/node_modules/grunt-contrib-uglify/node_modules/uglify-js/lib/parse.js:1327:19)' }
data:    app.js:8674 - >> Uglifying source .tmp/public/concat/production.js failed.
data:    app.js:8674 - Warning: Uglification failed.
data:    app.js:8674 - Name expected. 
data:    app.js:8674 - Line 5976 in .tmp/public/concat/production.js
````

Changing to development results in Sails lifting without issue.

I've tried making sure all the deps are installed with `npm install`, and also checked the permissions are correct on `.tmp/`.

Running `sails-0.10.5.`

**Comment by lsvx** [comment]
Any progress on this? I am seeing exactly the same error. Lifting sails with any other environment variable setting works fine, but sails won't lift in production.

**Comment by lsvx** [comment]
I finally just figured it out after days: I needed to explicitly have `grunt` listed as a dependency of my project. Weird that this is necessary in production but not other environments.

**Comment by CWyrtzen** [comment]
does @lsvx comment resolve the question? Is this still an issue? Can we close this up?

**Comment by nheinbaugh** [comment]
I am having the same issue. I don't think it should be closed. I think it should be fixed!

**Comment by wvezey** [comment]
I agree with nheinbaugh, especially since we are working on the same project. We have grunt listed as a dependency in the package.json file, but are getting grunt-related errors, similar to the one posted above.  We are getting grunt errors related to the assets compiling and linking tasks. When we remove those from the tasks > register > defaults.js file, Sails lifts normally.

**Comment by sailsbot** [comment]
Thanks for posting, @wyqydsyq.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by Cris87** [comment]
Hello everyone, I am unfortunately seeing this same error, @lsvx if it's not too much trouble, could you please tell me how to explicitly have grunt listed as a dependency of my project? Thank you so much.

**Comment by wyqydsyq** [comment]
@Cris87 `npm install grunt --save` should work

**Comment by Cris87** [comment]
Thank you for the quick response @wyqydsyq  unfortunately it did not work. I've tried everything else, changing the permissions in .tmp, npm install, checking for errors in all other assets. It runs smoothly when I do "sails lift" but when I run "node app.js --prod" I always get the same error:

> node app.js --prod      
Warning: connect.session() MemoryStore is not
designed for a production environment, as it will leak
memory, and will not scale past a single process.
** Grunt :: An error occurred. **
error: 
------------------------------------------------------------------------
Aborted due to warnings.
Running "uglify:dist" (uglify) task
JS_Parse_Error {
  message: 'Unexpected token name «i», expected punc «;»',
  filename: '../concat/production.js',
  line: 46920,
  col: 25,
  pos: 1639697,
  stack: 'Error\n    at new JS_Parse_Error (eval at <anonymous> (/usr/local/lib/node_modules/sails/node_modules/grunt-contrib-uglify/node_modules/uglify-js/tools/node.js:22:1), <anonymous>:1526:18)\n    at js_error (eval at <anonymous> (/usr/local/lib/node_modules/sails/node_modules/grunt-contrib-uglify/node_modules/uglify-js/tools/node.js:22:1), <anonymous>:1534:11)\n    at croak (eval at <anonymous> (/usr/local/lib/node_modules/sails/node_modules/grunt-contrib-uglify/node_modules/uglify-js/tools/node.js:22:1), <anonymous>:2026:9)\n    at token_error (eval at <anonymous> (/usr/local/lib/node_modules/sails/node_modules/grunt-contrib-uglify/node_modules/uglify-js/tools/node.js:22:1), <anonymous>:2034:9)\n    at expect_token (eval at <anonymous> (/usr/local/lib/node_modules/sails/node_modules/grunt-contrib-uglify/node_modules/uglify-js/tools/node.js:22:1), <anonymous>:2047:9)\n    at expect (eval at <anonymous> (/usr/local/lib/node_modules/sails/node_modules/grunt-contrib-uglify/node_modules/uglify-js/tools/node.js:22:1), <anonymous>:2050:36)\n    at regular_for (eval at <anonymous> (/usr/local/lib/node_modules/sails/node_modules/grunt-contrib-uglify/node_modules/uglify-js/tools/node.js:22:1), <anonymous>:2280:9)\n    at for_ (eval at <anonymous> (/usr/local/lib/node_modules/sails/node_modules/grunt-contrib-uglify/node_modules/uglify-js/tools/node.js:22:1), <anonymous>:2276:16)\n    at eval (eval at <anonymous> (/usr/local/lib/node_modules/sails/node_modules/grunt-contrib-uglify/node_modules/uglify-js/tools/node.js:22:1), <anonymous>:2158:24)\n    at eval (eval at <anonymous> (/usr/local/lib/node_modules/sails/node_modules/grunt-contrib-uglify/node_modules/uglify-js/tools/node.js:22:1), <anonymous>:2073:24)\n    at block_ (eval at <anonymous> (/usr/local/lib/node_modules/sails/node_modules/grunt-contrib-uglify/node_modules/uglify-js/tools/node.js:22:1), <anonymous>:2353:20)' }
>> Uglifying source .tmp/public/concat/production.js failed.
Warning: Uglification failed.
Unexpected token name «i», expected punc «;». 
Line 46920 in .tmp/public/concat/production.js
 
------------------------------------------------------------------------

error: Looks like a Grunt error occurred--
error: Please fix it, then **restart Sails** to continue running tasks (e.g. watching for changes in assets)
error: Or if you're stuck, check out the troubleshooting tips below.

error: Troubleshooting tips:
error: 
error:  *-> Are "grunt" and related grunt task modules installed locally?  Run `npm install` if you're not sure.
error: 
error:  *-> You might have a malformed LESS, SASS, CoffeeScript file, etc.
error: 
error:  *-> Or maybe you don't have permissions to access the `.tmp` directory?
error:      e.g., `/.tmp` ?
error: 
error:      If you think this might be the case, try running:
error:      sudo chown -R 501 /.tmp

Error: The hook `grunt` is taking too long to load.
Make sure it is triggering its `initialize()` callback, or else set `sails.config.grunt._hookTimeout to a higher value (currently 20000)
    at tooLong [as _onTimeout] (/usr/local/lib/node_modules/sails/lib/app/private/loadHooks.js:85:21)
    at Timer.listOnTimeout (timers.js:92:15) { [Error: The hook `grunt` is taking too long to load.
Make sure it is triggering its `initialize()` callback, or else set `sails.config.grunt._hookTimeout to a higher value (currently 20000)] code: 'E_HOOK_TIMEOUT' }
